### PR TITLE
[Bugfix][Spec Decode] Keep hidden-state extractor drafts at PP1

### PR DESCRIPTION
**Pull request by venkywonka**
## Purpose

Fix only the `extract_hidden_states` configuration branch: its last-stage-only extractor must use a draft parallel configuration with pipeline-parallel size one, rather than aliasing the target's parallel configuration. The target keeps its original pipeline-parallel size and object identity, and the extractor keeps the target tensor-parallel size and the existing factory's supported settings.

Use the existing `SpeculativeConfig.create_draft_parallel_config` factory without changing its implementation. Add five cases to the existing configuration test suite: target TP4/PP2, target TP8/PP1, rejection of an explicitly PP2 extractor, rejection of an invalid TP3 extractor, and rejection of an extractor configured as a PP2 target. The fixture writes only a local Llama configuration; it does not load weights or download a model.

This is an inherited extraction-only configuration fix, separate from MiniMax pipeline auxiliary-hidden-state support, ordinary EAGLE3 execution, autotuning, and any minimum performance stack. It does not relax `SupportsPP` or tensor-parallel divisibility validation.

### Duplicate-work check

Checked open PRs for hidden-state extraction, pipeline parallelism, and `create_draft_parallel_config`, including the overlapping configuration patches returned by those searches. None addresses the extractor branch aliasing the target parallel configuration. Dense-drafter expert-parallel inheritance changes are a separate issue and are not included here.

## Test Plan

The five added cases call the real configuration constructors and registry verification, without mocking pipeline support or bypassing validation. In a provisioned test environment, run:

```sh
HF_HUB_OFFLINE=1 TRANSFORMERS_OFFLINE=1 "$VLLM_TEST_PYTHON" -m pytest tests/test_config.py -k test_extract_hidden_states_ -q
```

`VLLM_TEST_PYTHON` denotes the Python interpreter in the existing vLLM test virtual environment.

## Test Result

- The reviewed production and test hunks apply unchanged to fresh upstream main. Hunk/context comparison, Python syntax checks, and comparison of the added fixture/test ASTs pass.
- Normal repository checks pass with `pre-commit run --files vllm/config/speculative.py tests/test_config.py`, including Ruff and the applicable configuration checks.
- **The five current-public-candidate regression tests now pass through normal repository imports: 5 passed, no failures, errors, or skips.** The same five test bodies against the full immediate pre-fix source produce **2 failed / 3 passed**, reproducing unsupported PP2 extraction and PP1 parallel-config aliasing. This is genuine imported pre-fix RED / candidate GREEN, not only the earlier source-fragment tests.
- The repository-test run used two CPUs and 8 GiB with no allocated GPUs, offline model access, no dependency installation, and CUDA initialization remaining false before and after pytest in its parent process. The expected pre-fix failures remain nonzero in scheduler records. Earlier missing-dependency collection attempts remain historical failures, not the current repository-test result.

## Standalone GPU validation — September 17, 2026 UTC

The current PR source now passes a source-isolated, real-weight Llama hidden-state extraction check on two SM120 GPUs, without #57197 or another PR. Dummy-weight startup passed first; the real-weight server then completed two HTTP 200 requests, each with 131 prompt tokens and 128 generated tokens.

Observed worker configuration confirms a TP1/PP2 target and TP1/PP1 extractor. Source-bound witnesses cover both requests' prefill and first decode: ordered auxiliary-state packing/relay across both pipeline stages, finite sampled values, and last-stage extractor consumption/cache correspondence. Both rank traces parse completely and contain actual CUDA kernel events. The model step exits successfully, post-exit checks find no owned model processes, assigned-GPU applications or service-port listener, and the allocation is released.

This closes the isolated-source intended-path GPU gap for this configuration fix. It reuses compatible native extensions and provides bounded inference/path coverage, not a fresh native build, completed-answer accuracy, performance comparison, ordinary EAGLE3 qualification or universal non-regression. The earlier failed private port probe remains a failed attempt; its real-socket regression was reproduced and repaired in the test runner only, with no production-source change. The existing five imported configuration tests and their pre-fix RED / candidate GREEN evidence remain unchanged.

AI assistance was used. The PR remains ready for human review; upstream CI still requires maintainer authorization, and the previously denied CI request is not repeated.

## GPU validation — September 17, 2026 UTC

The published configuration repair is exercised together with #57197 in two independent eight-SM120 MiniMax-M3 extraction controls. Observed runtime configuration records TP4/PP2 target with TP4/PP1 extractor, and TP8/PP1 target with TP8/PP1 extractor, both using `extract_hidden_states`.

The fresh controls each serve two HTTP 200 responses with 269 prompt tokens and 128 generated reasoning tokens per request; each repeat reuses 256 cached prompt tokens. All four PP1 and eight PP2 representative auxiliary-state witnesses pass. Profiler start and stop both return HTTP 200, with complete request/profile sessions of 227.03 seconds (PP1) and 121.21 seconds (PP2) within the unchanged 300-second limit. The supported stack-tracing option is observed as false; shape recording is retained.

Source/origin and post-exit owned-process/GPU-app/service-port checks pass. All eight rank traces per control pass complete gzip/JSON parsing and contain actual GPU kernel events and recorded shapes. Both model steps completed successfully and authoritative scheduler records confirm both allocations were released. Separate controller release/reconciliation/transport nonzero exits remain retained, and earlier profiler-stop failures remain failed. This is dependency-enabled configuration/inference coverage with reused native extensions, not an isolated current-head model run, fresh native build, full-answer accuracy, comparative performance, or universal non-regression claim.

AI assistance was used. This PR is ready for human review on the scoped evidence above. The existing upstream CI request requires maintainer authorization; no CI success or human approval is claimed, and the denied request is not repeated.

### CODE REVIEW

**Comment by venkywonka** [comment]
Repository-test validation update — September 17, 2026 UTC:

The five selected hidden-state extraction configuration tests now execute through normal repository imports on the current public candidate: **5 passed, 0 failed, 0 errors, 0 skipped**. The run used the full candidate source rather than transplanted implementation fragments.

The same five test bodies against the full pre-fix source give **2 failed and 3 passed**. The failures reproduce the PP2 unsupported-model configuration and PP1 parallel-config object aliasing; the candidate resolves both while retaining the invalid-configuration and model-capability checks. Only the regression tests were overlaid on the pre-fix source. This is observed pre-fix RED / candidate GREEN, not an inferred regression test.

Source and imported-module identities, test selection, raw pytest/JUnit outcomes and result hashes were checked. The allocation used two CPUs and 8 GiB with no allocated GPUs; Hugging Face access was offline, and PyTorch CUDA initialization remained false before and after pytest in its parent process. No dependency installation or model inference was performed. The scheduler retains a failed job status because the expected pre-fix failures were propagated, not rewritten as a successful job. The allocation and owned steps have terminated.

This closes the previously reported unexecuted repository-test gap. It does not claim GPU inference, production EAGLE3 qualification, upstream CI success or human approval; this separate configuration PR is now ready for human review based on the passing repository tests and independent scoped review. AI assistance was used.

**Comment by venkywonka** [comment]
GPU validation update — September 17, 2026 UTC:

The published configuration repair has now been exercised with #57197 in two independent eight-SM120 MiniMax hidden-state extraction controls: TP4/PP2 and TP8/PP1. Both loaded the real pretrained model and served two requests each. All four responses were HTTP 200 with 269 prompt tokens and 128 generated reasoning tokens, ending at the requested token limit; each repeat reported 256 cached prompt tokens.

This is positive dependency-enabled configuration/inference evidence, not an isolated-head or fresh-native-build qualification. The runs use the compatible configuration hunk with #57197 and retained prebuilt native extensions. The previously reported repository regression remains genuine pre-fix RED / candidate GREEN: two failures versus five passing selected tests.

Both GPU controls later timed out stopping profiling, and their complete auxiliary-state path assertions remain unqualified. Those failures are retained, not converted to full passes. Both owned allocations were released, with separate post-exit checks finding no owned workload processes, GPU applications or serving-port listeners. No completed-answer accuracy or comparative performance claim is made.

This PR remains ready for human review on its existing scoped evidence; upstream CI still requires a maintainer's authorization. AI assistance was used.

**Comment by venkywonka** [comment]
GPU validation update — September 17, 2026 UTC

The two fresh MiniMax-M3 extraction controls with #57197 now complete their original request/profile sessions, including successful profiler stop, without another configuration-source change:

- The observed TP4/PP2 target uses a TP4/PP1 extractor. The TP8/PP1 target uses a TP8/PP1 extractor. Both report the actual `extract_hidden_states` method.
- Two HTTP 200 responses per control, each with 269 prompt tokens and 128 generated reasoning tokens; each repeat reuses 256 cached prompt tokens. All four PP1 and eight PP2 representative auxiliary-state witnesses pass.
- Profiler start and stop both return HTTP 200. Complete request/profile sessions take 227.03 seconds and 121.21 seconds, respectively, within the original 300-second limit, with the supported stack-tracing option observed as false.
- Source/origin and post-exit owned-process/GPU-app/service-port checks pass. All eight traces per control are complete gzip/JSON exports with actual GPU kernel events and recorded shapes in every rank. Both model steps completed successfully and authoritative scheduler records confirm both allocations were released; separate controller release/reconciliation/transport nonzero exits remain retained.

This is dependency-enabled evidence for the published configuration repair with #57197 and reused native extensions, not an isolated current-head model run or fresh native build. Earlier profiler-stop failures remain failed. The prior imported repository regression remains genuine pre-fix RED (2 failed / 3 passed) and current-candidate GREEN (5 passed), with no test errors or skips.

The PR remains ready for human review on its scoped evidence. No full-answer accuracy, comparative performance, universal non-regression, or upstream CI success is claimed. The existing CI request requires maintainer authorization and is not being repeated. AI assistance was used.

**Comment by venkywonka** [comment]
Standalone GPU validation now passes on the current PR source, without #57197 or another PR:

- Two SM120 GPUs; dummy startup followed by two real-weight Llama requests, both HTTP 200, each generating 128 tokens.
- Actual TP1/PP2 target and TP1/PP1 extractor. Eight source-bound witnesses cover both requests' prefill/first decode, ordered PP auxiliary-state relay, finite samples and downstream extractor/cache consumption.
- Both complete rank traces contain CUDA kernel work. The model step succeeds; owned-process, GPU-app and service-port cleanup are verified and the allocation is released. Teardown warnings remain retained.

No production change was needed. This is bounded intended-path inference validation, not accuracy/performance or universal non-regression evidence; native extensions were reused. Existing configuration RED/GREEN tests remain valid. Ready for human review; CI still needs maintainer authorization. AI assistance was used.

**Comment by venkywonka** [comment]
Closing as superseded by #57197, whose published branch now includes the extraction-only draft PP1 configuration support, the original history and the existing regression tests. The replacement branch and complete changed-file union have been verified.

Prior GPU validation remains documented here with its original scope; it is not being relabeled as validation of the new combined head. The replacement stays draft pending the minimal combined-source checks, renewed review and required CI. Please continue review and fixes on #57197 rather than landing both copies.
